src/coap_client_zephyr.c

```diff
@@ -0,0 +1,1147 @@
+/*
+ * Copyright (c) 2022 Golioth, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <assert.h>
+#include <stdbool.h>
+#include <string.h>
+#include <netdb.h>      // struct addrinfo
+#include <sys/param.h>  // MIN
+#include <sys/select.h>
+#include <time.h>
+#include <golioth/golioth_debug.h>
+#include <golioth/golioth_sys.h>
+#include "coap_client.h"
+#include "golioth_util.h"
+#include "mbox.h"
+
+#include "coap_client_zephyr.h"
+#include "pathv.h"
+#include "zephyr_coap_req.h"
+#include "zephyr_coap_utils.h"
+
+#include <sys/eventfd.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/net/tls_credentials.h>
+
+#include <mbedtls/ssl_ciphersuites.h>
+#include <golioth_ciphersuites.h>
+
+LOG_MODULE_REGISTER(golioth_coap_client_zephyr);
+
+#define GOLIOTH_MAX_IDENTITY_LEN    32
+#define GOLIOTH_EMPTY_PACKET_LEN    (16 + GOLIOTH_MAX_IDENTITY_LEN)
+
+enum {
+    FLAG_RECONNECT,
+    FLAG_STOP_CLIENT,
+};
+
+#define PING_INTERVAL       (CONFIG_GOLIOTH_COAP_CLIENT_PING_INTERVAL_SEC * 1000)
+#define RECV_TIMEOUT        (CONFIG_GOLIOTH_COAP_CLIENT_RX_TIMEOUT_SEC * 1000)
+
+enum pollfd_type {
+    POLLFD_EVENT,
+    POLLFD_SOCKET,
+    POLLFD_MBOX,
+    NUM_POLLFDS,
+};
+
+static atomic_t flags;
+
+static uint8_t rx_buffer[CONFIG_GOLIOTH_COAP_CLIENT_RX_BUF_SIZE];
+
+static struct zsock_pollfd fds[NUM_POLLFDS];
+
+static sec_tag_t sec_tag_list[] = {
+    CONFIG_GOLIOTH_COAP_CLIENT_CREDENTIALS_TAG,
+};
+
+/* Use mbedTLS macros which are IANA ciphersuite names prepended with MBEDTLS_ */
+#define GOLIOTH_CIPHERSUITE_ENTRY(x) _CONCAT(MBEDTLS_, x)
+
+static int golioth_ciphersuites[] = {
+    FOR_EACH_NONEMPTY_TERM(GOLIOTH_CIPHERSUITE_ENTRY, (,), GOLIOTH_CIPHERSUITES)
+};
+
+static bool _initialized;
+
+/* Golioth instance */
+struct golioth_client _golioth_system_client;
+
+static inline void client_notify_timeout(void) {
+    eventfd_write(fds[POLLFD_EVENT].fd, 1);
+}
+
+static void golioth_system_client_wakeup(struct golioth_client* client) {
+    eventfd_write(fds[POLLFD_EVENT].fd, 1);
+}
+
+static void eventfd_timeout_handle(struct k_work *work) {
+    client_notify_timeout();
+}
+
+static K_WORK_DELAYABLE_DEFINE(eventfd_timeout, eventfd_timeout_handle);
+
+static enum coap_content_format golioth_content_type_to_coap_format(enum golioth_content_type content_type) {
+    switch (content_type)
+    {
+        case GOLIOTH_CONTENT_TYPE_JSON:
+            return COAP_CONTENT_FORMAT_APP_JSON;
+        case GOLIOTH_CONTENT_TYPE_CBOR:
+            return COAP_CONTENT_FORMAT_APP_CBOR;
+        default:
+            assert(0);
+            return COAP_CONTENT_FORMAT_APP_OCTET_STREAM;
+    }
+}
+
+static int golioth_coap_empty(golioth_coap_request_msg_t* req) {
```

Comment:
Do we need both this function and `__golioth_send_empty_coap()`? It seems like we should be able to replace `golioth_coap_empty(req)` with `__golioth_send_empty_coap(req->client)`